arrayUtilities: init

ConnorBaker opened the pull request:
There are patterns throughout the shell scripts in Nixpkgs (particularly those involving ELF files and CUDA) which can/should be componentized and made into a rigorous, tested library of functions. Given the push for `__structuredAttrs`, it makes sense to offer high(er)-level array utilities rather than rely on each individual's knowledge of Bash. Promoting code-reuse in this way avoids disparate re-implementations of common patterns which introducing bugs, reduces line count, and makes intent unambiguous. 

This PR is the beginning of such re-usable library of Bash functions. Currently, it includes:

- `isDeclaredArray`: a function which tests whether a nameref refers to a declared, indexed array
- `isDeclaredMap`: a function which tests whether a nameref refers to a declared, associative array
- `sortArray`: a function which lexicographically sorts the contents of an indexed array
- `getSortedMapKeys`: a function which populates an indexed array with the keys of an associative array, sorted in lexicographic order

Functions in (and to be added to) this library are largely written in destination-passing style, where the final arguments are the outputs in which the result of calling the function will be stored. (This is enabled by heavy use of namerefs.) Constraints on whether arguments may alias each other are documented for each relevant function.

The bulk of this PR is dedicated to testing these bash functions (generated with `cloc pkgs/build-support/setup-hooks/arrayUtilities --md`):

Language|files|blank|comment|code
:-------|-------:|-------:|-------:|-------:
Nix|8|84|12|981
Bourne Again Shell|4|15|47|49
--------|--------|--------|--------|--------
SUM:|12|99|59|1030

The second commit of this PR shows where such functionality would have been useful in implementing `testers.testEqualArrayOrMap`.

#### `isDeclaredArray` and `isDeclaredMap`

These two functions underpin the library and allow testing the type of variable to which a nameref refers.

#### `sortArray`

This function performs a stable-sort of an indexed array passed by reference, storing the results to the second argument, which is also an indexed array passed by reference.

Two points of note in the implementation:

1. When the number of directives in the format string does not match the number of arguments provided to `printf`, the format string is applied to each argument. In this way, we produce a NULL-delimited list of values for consumption by `sort`.
2. Since the locale changes the behavior of `sort`, we specify it in a sub-shell to avoid polluting the enclosing environment.

#### `getSortedMapKeys`

This function accepts an associative array by reference and stores the lexicographically sorted keys in the second argument, an indexed array passed by reference. Since order of traversal of associative arrays in Bash is not guaranteed to be consistent and Nix pursues reproducibility, being able to iterate over the values of a map in a deterministic fashion is a boon -- especially for test scripts. 

## Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [x] x86_64-darwin
  - [x] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [x] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#linking-nixos-module-tests-to-a-package) to the relevant packages
- [x] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [25.05 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2505.section.md) (or backporting [24.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2411.section.md) and [25.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2505.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

doronbehar commented:
> There's common functionality throughout Nixpkgs setup hooks which can/should be componentized and made into a rigorous, tested library of functions.
> 
> With the push for `__structuredAttrs`, it makes sense to offer high(er)-level array utilities.

Could you give please a somewhat simple and self contained example of existing Nixpkgs code that would gain from the changes scattered in these different PRs?

ConnorBaker committed:
testers.testEqualArrayOrMap: use arrayUtilities where possible

Signed-off-by: Connor Baker <ConnorBaker01@gmail.com>

ambroisie reviewed (commented):
This is pretty cool :-).

---

I'm kind of failing to see the value of `occursOnlyOrAfterInArray` compared to a two-step check of:

1. `elem1` in array
2. `elem1` strictly occurs after `elem2` (fails if `elem2` is not present).

Same for `occursOnlyOrBeforeInArray`.

ConnorBaker commented on pkgs/build-support/setup-hooks/arrayUtilities/arrayDifference/arrayDifference.bash:
```diff
@@ -0,0 +1,63 @@
+# shellcheck shell=bash
+
+# arrayDifference
+# Computes the difference of two arrays.
```
Good idea -- I updated the description to make that more clear.

ConnorBaker commented on pkgs/build-support/setup-hooks/arrayUtilities/arraysAreEqual/tests.nix:
```diff
@@ -0,0 +1,88 @@
+      "ERROR: genericBuild: arraysAreEqual returned unsuccessfully"
+    ];
+  };
+}
```
Added tests -- good suggestion!

ConnorBaker commented on pkgs/build-support/setup-hooks/arrayUtilities/getElfFiles/getElfFiles.bash:
```diff
@@ -0,0 +1,47 @@
+    fi
+
+    # NOTE: Since the input is sorted, the output is sorted by virtue of us iterating over it in order.
+    nixInfoLog "including $file"
```
Agreed, removed in latest

ConnorBaker commented:
Rebased on master and force-pushed -- thank you for the review @ambroisie!

I removed `occursOnlyOrAfterInArray` and `occursOnlyOrBeforeInArray` because they are trivial and I'm no longer using them in `cuda-packages`.

ConnorBaker commented:
Woah rebuilds! Working on reducing that back to a reasonable number ;)

philiptaron commented:
I've requested a review from @tie and @wolfgangwalther for their bash expertise.

philiptaron commented:
At a first glance, this all looks sane. Thanks for the extensive sets of tests. That made the dense and feature-rich bash code a lot less stomach churn inducing.

ConnorBaker committed:
arrayUtilities: init

Signed-off-by: Connor Baker <ConnorBaker01@gmail.com>

ConnorBaker commented:
I rebased on `master`, greatly reduced the number of functions included in this PR (to just those required for the rewrite of `testers.testEqualArrayOrMap`), updated the PR description, and included a breakdown of the line count (most lines belong to tests).

wolfgangwalther reviewed (commented):
Thanks for reducing scope, that helps a lot!

ConnorBaker commented on pkgs/build-support/setup-hooks/arrayUtilities/isDeclaredArray/package.nix:
```diff
@@ -0,0 +1,13 @@
+  passthru.tests = callPackages ./tests.nix { };
+  meta = {
+    description = "Tests if an array is declared";
+    maintainers = [ lib.maintainers.connorbaker ];
```
Resolved in latest

ConnorBaker commented:
Addressed some comments, rebased, and force-pushed.

Generally, I:

- Want to be explicit about variable scoping
- Want functions to check their arguments 
- Want functions to be restricted to operating on inputs which were declared with `declare` or `local`
- Want to keep using output-passing style

ConnorBaker commented on pkgs/build-support/setup-hooks/arrayUtilities/sortArray/sortArray.bash:
```diff
@@ -0,0 +1,43 @@
+#
+# Arguments:
+# - inputArrRef: a reference to an array (not mutated, may alias outputArrRef)
+# - outputArrRef: a reference to an array (mutated only by appending, may alias inputArrRef)
```
Fair enough -- I added it only in so far as to make clear the properties the function has, I'm okay with removing it to say that the output is fully replaced by the resulting sorted array.

ConnorBaker commented:
Sorry for the delay in getting back to your comments, I'll try to address those today.

ConnorBaker commented on pkgs/build-support/setup-hooks/arrayUtilities/getMapKeys/getMapKeys.bash:
```diff
@@ -0,0 +1,40 @@
+
+# getMapKeys
+# Appends a sorted array of the keys of the input associative array to the output arrray.
+#
```
I've changed `getMapKeys` to be non-appending just like `sortArray`.

The reasoning for using `getMapKeys` is precisely because the results keys are sorted -- the order of iteration on associative maps isn't guaranteed to be consistent, and when/since these scripts can be used by testers, stomping out non-determinism is important to me. As an added benefit it checks that the arguments provided are of correct type (unlike native parameter expansion which will accept expansions of strings).

wolfgangwalther reviewed (approved):
I did not go through the extensive test-suite in detail, but they are a great base and give me confidence that the implementation is solid. I did test whether some sorting tests would fail without sorting behavior - they did, cool.

LGTM in principle.

ConnorBaker commented on pkgs/build-support/setup-hooks/arrayUtilities/isDeclaredArray/isDeclaredArray.bash:
```diff
@@ -0,0 +1,14 @@
+# - inputArrayRef: a reference to an array (not mutated)
+#
+# Returns 0 if the array is declared, 1 otherwise.
+isDeclaredArray() {
+  # NOTE: We must dereference the name ref to get the type of the underlying variable.
+  # shellcheck disable=SC2034
+  local -nr inputArrayRef="$1" && [[ ${!inputArrayRef@a} =~ a ]]
+}
```
I really like that these functions are separate so I can use them outside of things providing `setup.sh` and have an easy way to test them. I'm not opposed to the ol' copy/paste or including the files in `mkDerivation`'s `nativeBuildInputs`... what are your thoughts? Although maybe copy-pasting would be better so any changes to these functions don't cause a rebuild of `stdenv`.
